### PR TITLE
(GH-1729) Add option to skip clean up of tmpdir

### PR DESCRIPTION
**Pull request by nicklewis**
!feature

* **`--no-cleanup` option to leave behind temporary files**
  ([#1729](https://github.com/puppetlabs/bolt/issues/1729))

  The `--no-cleanup` flag or `cleanup: false` transport option can now
  be set to instruct Bolt not to clean up on a target after it's
  finished. This is useful for debugging what Bolt is doing on a system.

### CODE REVIEW

File: lib/bolt/config.rb [hunk 1]
```diff
@@ -36,6 +36,7 @@ class Config
 
     OPTIONS = {
       "apply_settings"           => "A map of Puppet settings to use when applying Puppet code",
+      "cleanup"                  => "Whether to clean up temporary files created on targets",
```
Comment (beechtom) [inline lib/bolt/config.rb:39]: The changes in `Bolt::Config` look like they're not needed - I'm guessing this is from before deciding it'd be a transport-level option?

Comment (nicklewis) [reply lib/bolt/config.rb:39]: Whoops yeah. I'll remove that.

File: lib/bolt/config/transport/docker.rb [hunk 2]
```diff
@@ -8,6 +8,8 @@ class Config
     module Transport
       class Docker < Base
         OPTIONS = {
+          "cleanup"       => { type: TrueClass,
+                               desc: "Whether to clean up temporary files created on targets" },
```
Comment (beechtom) [inline lib/bolt/config/transport/docker.rb:12]:
```suggestion
                               desc: "Whether to clean up temporary files created on targets." },
```

File: lib/bolt/config/transport/local.rb [hunk 3]
```diff
@@ -8,6 +8,8 @@ class Config
     module Transport
       class Local < Base
         OPTIONS = {
+          "cleanup"         => { type: TrueClass,
+                                 desc: "Whether to clean up temporary files created on targets" },
```
Comment (beechtom) [inline lib/bolt/config/transport/local.rb:12]:
```suggestion
                                 desc: "Whether to clean up temporary files created on targets." },
```

File: lib/bolt/config/transport/local.rb [hunk 4]
```diff
@@ -36,6 +38,8 @@ class Local < Base
         }.freeze
 
         WINDOWS_OPTIONS = {
+          "cleanup"      => { type: TrueClass,
+                              desc: "Whether to clean up temporary files created on targets" },
```
Comment (beechtom) [inline lib/bolt/config/transport/local.rb:42]:
```suggestion
                              desc: "Whether to clean up temporary files created on targets." },
```

File: lib/bolt/config/transport/ssh.rb [hunk 5]
```diff
@@ -9,6 +9,8 @@ module Transport
       class SSH < Base
         LOGIN_SHELLS = %w[sh bash zsh dash ksh powershell].freeze
         OPTIONS = {
+          "cleanup"            => { type: TrueClass,
+                                    desc: "Whether to clean up temporary files created on targets" },
```
Comment (beechtom) [inline lib/bolt/config/transport/ssh.rb:13]:
```suggestion
                                    desc: "Whether to clean up temporary files created on targets." },
```

File: lib/bolt/config/transport/winrm.rb [hunk 6]
```diff
@@ -12,6 +12,8 @@ class WinRM < Base
                                  desc: "Force basic authentication. This option is only available when using SSL." },
           "cacert"          => { type: String,
                                  desc: "The path to the CA certificate." },
+          "cleanup"         => { type: TrueClass,
+                                 desc: "Whether to clean up temporary files created on targets" },
```
Comment (beechtom) [inline lib/bolt/config/transport/winrm.rb:16]:
```suggestion
                                 desc: "Whether to clean up temporary files created on targets." },
```

**Comment by logicminds** [comment]
Awesome!

**Pull request merged**